### PR TITLE
AUC and Logloss

### PR DESCRIPTION
**Pull request by yueguoguo**
### Description
Discussed with @Leavingseason and added
* AUC metric
* Logloss metric

### Related Issues
#445 

### Checklist:

- [x] My code follows the code style of this project, as detailed in our [contribution guidelines](../CONTRIBUTING.md).
- [x] I have added tests.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by yueguoguo** [comment]
Updates in the notebooks will be given in another PR for the convenience of review.

File: reco_utils/evaluation/python_evaluation.py [hunk 1]
```diff
@@ -591,6 +593,101 @@ def map_at_k(
...
+        col_prediction (str): column name for prediction.
+
+    Return:
+        float: auc_score (min=0, max=1).
+    """
+    rating_true_pred = merge_rating_true_pred(
+        rating_true, rating_pred, col_user, col_item, col_rating, col_prediction
+    )
+    auc_score = roc_auc_score(
+        np.asarray(rating_true_pred[DEFAULT_RATING_COL]),
```
Comment (miguelgfierro) [inline reco_utils/evaluation/python_evaluation.py]: not 100%, but I would say that `rating_true_pred[DEFAULT_RATING_COL].values` is faster

File: reco_utils/evaluation/python_evaluation.py [hunk 2]
```diff
@@ -591,6 +593,101 @@ def map_at_k(
...
+    rating_true_pred = merge_rating_true_pred(
+        rating_true, rating_pred, col_user, col_item, col_rating, col_prediction
+    )
+
+    # To make sure that there is no N/A in the logloss metric calculation.
+    rating_true_pred[PREDICTION_COL] = (
+        rating_true_pred[PREDICTION_COL].apply(lambda x: max(min(x, 1.0 - 10e-12), 10e-12))
+    )
+
+    log_loss_score = log_loss(
```
Comment (miguelgfierro) [inline reco_utils/evaluation/python_evaluation.py]: same comment as before

File: tests/unit/test_python_evaluation.py [hunk 3]
```diff
@@ -38,75 +41,89 @@ def target_metrics(scope="module"):
...
-                9,
-                8,
-                7,
-                6,
-                5,
-            ],
-        }
-    )
-    return rating_true, rating_pred, rating_nohit
+    def _generate_python_data(binary_rating=False):
```
Comment (miguelgfierro) [inline tests/unit/test_python_evaluation.py]: don't understand why this is needed, instead of directly returning the ratings

File: tests/unit/test_python_evaluation.py [hunk 4]
```diff
@@ -38,75 +41,89 @@ def target_metrics(scope="module"):
...
+            # Normalize the prediction.
+            rating_pred[PREDICTION_COL] = minmax_scale(rating_pred[PREDICTION_COL].astype(float))
+
+        return rating_true, rating_pred, rating_nohit
+    return _generate_python_data
 
 
 def test_column_dtypes_match(python_data):
-    rating_true, rating_pred, _ = python_data
+    rating_true, rating_pred, _ = python_data()
```
Comment (miguelgfierro) [inline tests/unit/test_python_evaluation.py]: here we are not using python_data as fixture but as function. The correct way to add it would be just `python_data`.  Here more about fixtures https://docs.pytest.org/en/latest/fixture.html

File: tests/unit/test_python_evaluation.py (hunk 3 above)
Comment (yueguoguo) [reply tests/unit/test_python_evaluation.py]: Because a function with argument `binary_rating` is returned from the fixture to allow testing different types of metrics, rating metric and binary prediction metric, respectively.

File: tests/unit/test_python_evaluation.py (hunk 4 above)
Comment (yueguoguo) [reply tests/unit/test_python_evaluation.py]: To pass a function, then I can use the parameter. See comment above.

File: tests/unit/test_python_evaluation.py (hunk 3 above)
Comment (miguelgfierro) [reply tests/unit/test_python_evaluation.py]: yeah, but then we are  using it as fixtures. Let me ping you offline

Comment (yueguoguo) [reply tests/unit/test_python_evaluation.py]: See https://alysivji.github.io/pytest-fixures-with-function-arguments.html

Comment (miguelgfierro) [reply tests/unit/test_python_evaluation.py]: oh I see what you are doing, sorry I didn't understand it initially. Yeah, you are right, this is correct. I think it is because I didn't see the argument. 
Would it make sense to set explicitly `binary_rating` in all the instances to make it clearer?

Comment (yueguoguo) [reply tests/unit/test_python_evaluation.py]: @miguelgfierro great point - will do that.

File: reco_utils/evaluation/python_evaluation.py [hunk 5]
```diff
@@ -591,6 +593,100 @@ def map_at_k(
...
+    rating_true_pred = merge_rating_true_pred(
+        rating_true, rating_pred, col_user, col_item, col_rating, col_prediction
+    )
+    auc_score = roc_auc_score(
+        np.asarray(rating_true_pred[DEFAULT_RATING_COL].values),
+        np.asarray(rating_true_pred[PREDICTION_COL].values)
+    )
+
+    # Round the AUC score to four digits.
+    auc_score = round(auc_score, 4)
```
Comment (anargyri) [inline reco_utils/evaluation/python_evaluation.py]: Sorry to be a bit late to the PR! 
This is a magic number. Why do we need to round?

File: reco_utils/evaluation/python_evaluation.py [hunk 6]
```diff
@@ -591,6 +593,100 @@ def map_at_k(
...
+        rating_true_pred[PREDICTION_COL].apply(lambda x: max(min(x, 1.0 - 10e-12), 10e-12))
+    )
+
+    log_loss_score = log_loss(
+        np.asarray(rating_true_pred[DEFAULT_RATING_COL].values),
+        np.asarray(rating_true_pred[PREDICTION_COL].values)
+    )
+
+    # Round the AUC score to four digits.
+    log_loss_score = round(log_loss_score, 4)
```
Comment (anargyri) [inline reco_utils/evaluation/python_evaluation.py]: Same here, this is a magic number.

File: reco_utils/evaluation/python_evaluation.py [hunk 7]
```diff
@@ -591,6 +593,100 @@ def map_at_k(
...
+    Return:
+        float: log_loss_score (min=-\inf, max=\inf).
+    """
+    rating_true_pred = merge_rating_true_pred(
+        rating_true, rating_pred, col_user, col_item, col_rating, col_prediction
+    )
+
+    # To make sure that there is no N/A in the logloss metric calculation.
+    rating_true_pred[PREDICTION_COL] = (
+        rating_true_pred[PREDICTION_COL].apply(lambda x: max(min(x, 1.0 - 10e-12), 10e-12))
```
Comment (anargyri) [inline reco_utils/evaluation/python_evaluation.py]: Again this is a magic number. Btw 10e-12 is larger than 1e-12.
You could also use `numpy.clip`
https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.clip.html

File: reco_utils/evaluation/python_evaluation.py [hunk 8]
```diff
@@ -591,6 +593,100 @@ def map_at_k(
...
+        col_prediction (str): column name for prediction.
+
+    Return:
+        float: auc_score (min=0, max=1).
+    """
+    rating_true_pred = merge_rating_true_pred(
+        rating_true, rating_pred, col_user, col_item, col_rating, col_prediction
+    )
+    auc_score = roc_auc_score(
+        np.asarray(rating_true_pred[DEFAULT_RATING_COL].values),
```
Comment (anargyri) [inline reco_utils/evaluation/python_evaluation.py]: `.values` returns already a numpy array.

File: reco_utils/evaluation/python_evaluation.py [hunk 9]
```diff
@@ -591,6 +593,100 @@ def map_at_k(
...
+        rating_true, rating_pred, col_user, col_item, col_rating, col_prediction
+    )
+
+    # To make sure that there is no N/A in the logloss metric calculation.
+    rating_true_pred[PREDICTION_COL] = (
+        rating_true_pred[PREDICTION_COL].apply(lambda x: max(min(x, 1.0 - 10e-12), 10e-12))
+    )
+
+    log_loss_score = log_loss(
+        np.asarray(rating_true_pred[DEFAULT_RATING_COL].values),
```
Comment (anargyri) [inline reco_utils/evaluation/python_evaluation.py]: Same about `.values`

File: reco_utils/evaluation/python_evaluation.py (hunk 7 above)
Comment (anargyri) [reply reco_utils/evaluation/python_evaluation.py]: There are redundant parentheses.

File: tests/unit/test_python_evaluation.py [hunk 10]
```diff
@@ -38,75 +41,89 @@ def target_metrics(scope="module"):
...
+                    7,
+                    6,
+                    5,
+                ],
+            }
+        )
+
+        if binary_rating:
+            # Convert to binary case.
+            rating_true[DEFAULT_RATING_COL] = (
```
Comment (anargyri) [inline tests/unit/test_python_evaluation.py]: No need to have parentheses.

File: tests/unit/test_python_evaluation.py [hunk 11]
```diff
@@ -38,75 +41,89 @@ def target_metrics(scope="module"):
...
+                rating_true[DEFAULT_RATING_COL].apply(lambda x: 1.0 if x >= 3 else 0.0)
```
Comment (anargyri) [inline tests/unit/test_python_evaluation.py]: Can use `numpy.clip`.

File: tests/unit/test_python_evaluation.py [hunk 12]
```diff
@@ -307,8 +324,42 @@ def test_python_recall(python_data, target_metrics):
...
 
 
+def test_python_auc(python_data, target_metrics):
+    rating_true, rating_pred, _ = python_data(binary_rating=True)
+
+    assert auc(
+        rating_true=rating_true,
+        rating_pred=rating_true,
+        col_prediction=DEFAULT_RATING_COL
+    ) == pytest.approx(1.0, 0.1)
```
Comment (anargyri) [inline tests/unit/test_python_evaluation.py]: The tolerance for this should be low. Also define a constant.

File: tests/unit/test_python_evaluation.py [hunk 13]
```diff
@@ -307,8 +324,42 @@ def test_python_recall(python_data, target_metrics):
...
+
+
+def test_python_logloss(python_data, target_metrics):
+    rating_true, rating_pred, _ = python_data(binary_rating=True)
+
+    assert logloss(
+        rating_true=rating_true,
+        rating_pred=rating_true,
+        col_prediction=DEFAULT_RATING_COL
+    ) == pytest.approx(0, 0.1)
```
Comment (anargyri) [inline tests/unit/test_python_evaluation.py]: Again magic number.

File: reco_utils/evaluation/python_evaluation.py (hunk 5 above)
Comment (yueguoguo) [reply reco_utils/evaluation/python_evaluation.py]: Yea it wrapped the codes in deeprec implementation. Removed rounding for consistency.

File: reco_utils/evaluation/python_evaluation.py (hunk 7 above)
Comment (yueguoguo) [reply reco_utils/evaluation/python_evaluation.py]: Ah the parentheses is generated from `black` but yea we can remove it.

Comment (yueguoguo) [reply reco_utils/evaluation/python_evaluation.py]: Even clip is not needed - this is handled with the `log_loss` function in `sklearn` directly. The default value is 1e-15. https://scikit-learn.org/stable/modules/generated/sklearn.metrics.log_loss.html

File: tests/unit/test_python_evaluation.py (hunk 10 above)
Comment (yueguoguo) [reply tests/unit/test_python_evaluation.py]: This was because of `black` but I removed them anyway.
